adjust lcmp overview (#1)
darwinia-network/darwinia-docs: docs/lcmp-overview.md

# LCMP Overview

LCMP is Darwinia’s light-client cross-chain messaging protocol(former name: Darwinia bridge messages protocol). It allows one blockchain to send messages to another blockchain as long as they have established messaging channels with each other. 

LCMP is designed to be a general purpose protocol that means LCMP knows nothing about the DApps running on it. 

## Design Principles

**Guaranteed delivery:**

> When a transaction has been successfully committed on the source chain, it will arrive the target chain as long as both chains are producing blocks.

Reviewer: ```suggestion
When a cross-chain message has been successfully committed on the source chain, it will arrive the target chain as long as both chains are producing blocks.
```

> When a cross-chain message has been successfully committed on the source chain, it will arrive the target chain as long as both chains are producing blocks.

Reviewer 2: 是不是可以改成“When a cross-chain message has been successfully emitted from the source chain, it will arrive at the target chain”

Author: "emit" is used with events in our docs.

> When a cross-chain message has been successfully committed on the source chain, it will arrive the target chain as long as both chains are producing blocks.

Author: be emiited还是be committed? event用的是emit

> When a cross-chain message has been successfully committed on the source chain, it will arrive the target chain as long as both chains are producing blocks.

Reviewer 3: 我也不确定，不知道这里的原意是想表达什么，看 Carl 提议 emit 所以我就顺应了

**Trustless interoperability:**

> Users does not have to trust an individual or party to act honestly. the smart contracts will ensure transactions won’t be able to be exploited by bad actors as long as the consensus of the source chain and the target chain is secure.

Reviewer: ```suggestion
Users does not have to trust an individual or party to act honestly. Light client model will ensure transactions won’t be able to be exploited by bad actors as long as the consensus of the source chain and the target chain is secure.
```

> Users does not have to trust an individual or party to act honestly. Light client model will ensure transactions won’t be able to be exploited by bad actors as long as the consensus of the source chain and the target chain is secure.

Reviewer 2: User do not

**Generalized message**

> Message sent across blockchains can be encoded and decoded in any way. The message structure is flexible while removing the complexity of building chain-specific integrations.  

Reviewer: ```suggestion
Message sent across blockchains can be encoded and decoded in arbitrary way. The message structure is flexible while removing the complexity of building chain-specific integrations.  
```

> For now, the message has implemented the call encoding in Substrate’s SCALE CODEC and Ethereum’s ABI format . But the structure of message is open to any format. In the near future, it will adopt Polkadot’s XCM format for more general.

Reviewer: ```suggestion
For now, the message format support encodings in Substrate’s SCALE CODEC and Ethereum’s ABI format . But the structure of message is open to any format. In the near future, it will adopt Polkadot’s XCM format to be more general.
```

> Message sent across blockchains can be encoded and decoded in arbitrary way. The message structure is flexible while removing the complexity of building chain-specific integrations.  

Reviewer 2: in an arbitrary way

> For now, the message format support encodings in Substrate’s SCALE CODEC and Ethereum’s ABI format . But the structure of message is open to any format. In the near future, it will adopt Polkadot’s XCM format to be more general.

Reviewer 2: Ethereum’s ABI format. 句号前多了空格

Author: good eyes

## High-level Overview

![high-level](/img/lcmpoverview-0.png)

LCMP is a cross-chain messaging protocol. LCMP has two layers in design. One is the truth layer which contains the light client. Another is the message layer used to send cross-chain messages. The light client in the target chain’s truth layer is used to verify messages from the source chain.

As the diagram shows, there are four components cross the source chain and the target chain.

> The DAPP of the `App Layer` calls send_message of `Message Layer` to start a cross-chain operation. When send_message is executed, the `Relayer` forwards the message to the target chain, and the `Truth Layer` in the target chain checks the message and dispatch the task carried by the message. After dispatch, the `Relayer` forwards the confirmation message back to the source chain. This process actually includes two cross-chain processes.

Reviewer: ```suggestion
The Dapp of the `App Layer` calls send_message of `Message Layer` to start a cross-chain operation. When send_message is executed, the `Relayer` forwards the message to the target chain, and the `Truth Layer` in the target chain checks the message and dispatch the task carried by the message. After dispatch, the `Relayer` forwards the confirmation message back to the source chain. This process actually includes two cross-chain processes.
```

`Dapp` is more common usage on internet if we google it

> The Dapp of the `App Layer` calls send_message of `Message Layer` to start a cross-chain operation. When send_message is executed, the `Relayer` forwards the message to the target chain, and the `Truth Layer` in the target chain checks the message and dispatch the task carried by the message. After dispatch, the `Relayer` forwards the confirmation message back to the source chain. This process actually includes two cross-chain processes.

Reviewer 2: checks the message and dispatches the task

> The Dapp of the `App Layer` calls send_message of `Message Layer` to start a cross-chain operation. When send_message is executed, the `Relayer` forwards the message to the target chain, and the `Truth Layer` in the target chain checks the message and dispatch the task carried by the message. After dispatch, the `Relayer` forwards the confirmation message back to the source chain. This process actually includes two cross-chain processes.

Author: 收藏

![high-level](/img/lcmpoverview-1.png)

## Message Flow

Here we have a more detailed diagram with the inner components which will help us to understand the cross-chain messaging flow.

![message-flow](/img/lcmpoverview-2.png)

### Message Lifecycle

In order to more easily understand the cross-chain flow, here we define the lifecycle of the message.

Once a user calls send_message, a message is processed and the "MessageAccepted" event is emitted. Its initial status is "`undelivered`".

When the Relayer calls the "receive_messages_proof" of the target chain, the message will eventually enter the "`delivered`" status and emit the "MessageDispathed" event.

When the Relayer calls the "receive_messages_delivery_proof" of the source chain, the message finally enters the "`confirmed`" status and emits the "MessageDelivered" event.

![lifecycle](/img/lcmpoverview-3.png)

### Flow

1. send_message
    

>     When a DAPP calls outbound's send_message function, outbound store the message and emits an "MessageAccepted" event. The message is now in the "undelivered" status.

Reviewer: ```suggestion
    When a Dapp calls outbound's send_message function, outbound will store the message and emit an "MessageAccepted" event. The message is now in the "undelivered" status.
```

>     When a Dapp calls outbound's send_message function, outbound will store the message and emit an "MessageAccepted" event. The message is now in the "undelivered" status.

Reviewer 2: a "MessageAccepted" event.

    
2. relay
    
    There are off-chain participants, called Relayers, who relay messages to blockchain on the other side. 
    

>     Relayers will take the "undelivered" messages and relay them. The "receive_messages_proof" will be executed in the inbound on the target chain(delivery transaction). Once the transaction is successful, the message will be considered "delivered".

Author: The relayers take the "undelivered" messages and relay them. Relayer then crafts "receive_messages_proof()" transaction (aka delivery transaction) for the messages module instance, deployed at the target chain. Once a transaction is mined, the message is considered "delivered".

ok?

>     After the message is sent, the Relayer need to send confirmation back to the source chain. There are two reasons for this.

Reviewer 2: the Relayer needs to

>     First, we limit the number of "`delivered`" but not yet unconfirmed messages in the inbound channel. Therefore, the target chain may stop accepting new messages until the relayer confirms some of them.

Reviewer: ```suggestion
    First, we limit the number of "`delivered`" messages not unconfirmed yet in the inbound channel. Therefore, the target chain may stop accepting new messages until the relayer confirms some of them.
```

>     Second, if a relayer wants to be rewarded, he must prove that he actually delivered the message. This proof can only be generated after a successful delivery transaction. Therefore, the relayer calls the "receive_messages_delivery_proof" (the confirmation transaction) of the outbound on the source chain. Once this transaction has been successful, the message will be considered "`confirmed`".

Author: 这边的原因我会替换成parity的句子

>     Second, if a Relayer wants to be rewarded, he must prove that he actually delivered the message. This proof can only be generated after a successful delivery transaction. Therefore, the Relayer calls the receive_messages_delivery_proof (the confirmation transaction) of the outbound on the source chain. Once this transaction has been successful, the message will be considered "`confirmed`".

Reviewer: ```suggestion
    Second, if a relayer wants to be rewarded, he must prove that he actually delivered the message. This proof can only be generated after a successful delivery transaction. Therefore, the relayer calls the "receive_messages_delivery_proof" (the confirmation transaction) of the outbound on the source chain. Once this transaction has been successful, the message will be considered "`confirmed`".
```

    
6. callback
    
    On the source chain, developers can provide a callback. In the callback, you can write post cross-chain actions.
    

>     It is important to note that the callback will be called even if the cross-chain operation failures, so this is also the place to write the failure logic.

Reviewer 3: ```suggestion
    Notice: the callback will be called whether the cross-chain operation succeeded or failed, so this is also the correct place to handle the errors.
```

> 7. reward

Reviewer 3: ```suggestion
7. Reward
```

>     Pay rewards to the Relayers.

Reviewer 3: ```suggestion
    Pay rewards to the relayers.
```

> 8. The Relayers include the latest "`confirmed`" message in the next receive_messages_proof transaction, proving that some messages have been confirmed. The "`confirmed`" state is the final state of the message. But there is still something to be done - the reward has been paid to the Relayer must be known to the target chain. Otherwise, it may reach the limit of unconfirmed messages on the target chain, which will stop accepting new messages. 

Reviewer 3: ```suggestion
8. The relayers include the latest "`confirmed`" message in the next receive_messages_proof transaction, proving that some messages have been confirmed. The "`confirmed`" state is the final state of the message. But there is still something that needs to be done - the reward that has been paid to the relayers must be known to the target chain. Otherwise, it may reach the limit of unconfirmed messages on the target chain, which will cause the target chain to stop accepting new messages. 
```

感觉这边表达有点绕，可能要改下句子

Author: 这块确实有点饶

![flow](/img/lcmpoverview-4.png)

> Note that the header relay of the light client is not showed in this diagram, but it is not a problem for  understanding.

Reviewer 2: is not shown

> Note that the header relay of the light client is not showed in this diagram, but it is not a problem for  understanding.

Reviewer 2: a problem for understanding. for后面多了一个空格

> Note that the header relay of the light client is not showed in this diagram, but it is not a problem for  understanding.

Reviewer 3: ```suggestion
Notice: the header relay of the light client is not shown in this diagram.
```
